src/c_api/c_api.cc

```diff
@@ -1018,36 +1018,73 @@ XGB_DLL int XGBoosterPredictFromDMatrix(BoosterHandle handle,
+                            Learner *learner,
+                            xgboost::PredictionType type,
+                            float missing,
+                            size_t n_rows, size_t n_cols,
+                            size_t iteration_begin, size_t iteration_end,
+                            bool strict_shape,
+                            xgboost::bst_ulong const **out_shape,
+                            xgboost::bst_ulong *out_dim, const float **out_result) {
+  HostDeviceVector<float>* p_predt { nullptr };
+  learner->InplacePredict(p_m, type, missing, &p_predt, iteration_begin, iteration_end);
+  CHECK(p_predt);
+  auto &shape = learner->GetThreadLocal().prediction_shape;
+  auto chunksize = n_rows == 0 ? 0 : p_predt->Size() / n_rows;
+  CalcPredictShape(strict_shape, type, n_rows, n_cols, chunksize, learner->Groups(),
+                   learner->BoostedRounds(), &shape, out_dim);
+  *out_result = dmlc::BeginPtr(p_predt->HostVector());
+  *out_shape = dmlc::BeginPtr(shape);
+}
+
 void InplacePredictImpl(std::shared_ptr<DMatrix> p_m, char const *c_json_config, Learner *learner,
                         xgboost::bst_ulong const **out_shape, xgboost::bst_ulong *out_dim,
                         const float **out_result) {
   xgboost_CHECK_C_ARG_PTR(c_json_config);
   auto config = Json::Load(StringView{c_json_config});
   CHECK_EQ(get<Integer const>(config["cache_id"]), 0) << "Cache ID is not supported yet";
 
-  HostDeviceVector<float> *p_predt{nullptr};
   auto type = PredictionType(RequiredArg<Integer>(config, "type", __func__));
   float missing = GetMissing(config);
-  learner->InplacePredict(p_m, type, missing, &p_predt,
-                          RequiredArg<Integer>(config, "iteration_begin", __func__),
-                          RequiredArg<Integer>(config, "iteration_end", __func__));
-  CHECK(p_predt);
-  auto &shape = learner->GetThreadLocal().prediction_shape;
   auto const &info = p_m->Info();
   auto n_samples = info.num_row_;
   auto n_features = info.num_col_;
-  auto chunksize = n_samples == 0 ? 0 : p_predt->Size() / n_samples;
+  int iteration_begin = get<Integer const>(config["iteration_begin"]);
+  int iteration_end   = get<Integer const>(config["iteration_end"]);
   bool strict_shape = RequiredArg<Boolean>(config, "strict_shape", __func__);
 
   xgboost_CHECK_C_ARG_PTR(out_dim);
-  CalcPredictShape(strict_shape, type, n_samples, n_features, chunksize, learner->Groups(),
-                   learner->BoostedRounds(), &shape, out_dim);
 
   xgboost_CHECK_C_ARG_PTR(out_result);
   xgboost_CHECK_C_ARG_PTR(out_shape);
 
-  *out_result = dmlc::BeginPtr(p_predt->HostVector());
-  *out_shape = dmlc::BeginPtr(shape);
+  InplacePredictImplCore(p_m, learner, type, missing, n_samples, n_features,
+                         iteration_begin, iteration_end, strict_shape, out_shape,
+                         out_dim, out_result);
+}
+
+XGB_DLL int XGBoosterInplacePredict(BoosterHandle handle,
```

Comment:
Hi, this is the same function as `XGBoosterPredictFromDense`, can we use it instead?

Reply:
Hi, I've been working on moving us to `XGBoosterPredictFromDense` but a direct usage would require us to use something similar to what is used in the python code, but from reading it it looks like the arrays that are passed through are created using methods from numpy, which there doesn't seem to be any native support or connection for these types in Java. Should we be building a custom byte stream to send in here to have all the metadata or would something like `com.google.gson.Gson` being added to the imports for Java be acceptable?

Reply:
Hi, I think one can build the JSON interface in xgboost.cpp using the `xgboost:Json` class.

Reply:
Hi, I've tried a few different ways of building the JSON interface locally, but the best I've managed failed the type check in PredictFromDense, so incorrect conversion. Could you point me to what methods in the Json class I should be using for converting a Java float array to the expected format in the api? I'm inexperienced with C++ and could use additional guidance on that side of the pull request.

Reply:
Apologies for the slow reply. I will look into your PR tomorrow and see if I can help.